zucchini/grading_manager.py

```diff
@@ -201,6 +207,118 @@ def breakdown(self, grader_name):
 
         return '{} -{}'.format(breakdown or 'Perfect!', grader_name)
 
+    def generate_gradelog(self):
+        """
+        Writes gradelog file to submission directory
+        """
+
+        gradelog_path = os.path.join(
+            self._submission.path, SUBMISSION_GRADELOG_FILE)
+
+        with open(gradelog_path, 'w') as f:
+
+            m, s = divmod(self._submission.seconds_late, 60)
+            h, m = divmod(m, 60)
+
+            f.write("student_name: \"%s\", hours_late: %s\n\n" % (
+                self._submission.student_name,
+                "%d:%02d:%02d" % (h, m, s)))
+
+            assignment_total = Fraction(0, 1)
+            assignment_out_of = Fraction(0, 1)
+
+            for component, component_grade in zip(self._assignment.components,
+                                                  self._component_grades):
+                if component_grade.is_broken():
+                    percentage_of_total = Fraction(
+                        component.weight, self._assignment.total_weight)
+                    component_points = self._left_pad(percentage_of_total)
+                    earned_score = self._left_pad(Fraction(0, 1))
+                    assignment_out_of += percentage_of_total
+                    f.write("(%s / %s) [FAIL] TOTAL for %s: %s\n\n" % (
+                        earned_score, component_points, component.name,
+                        component_grade.error))
+                else:
+                    component_total = Fraction(0, 1)
+                    component_out_of = Fraction(0, 1)
+
+                    for part, part_grade in zip(component.parts,
+                                                component_grade.part_grades):
+                        # calculate scores (deal with case where weight is 0)
+                        if 0 not in (component.weight, part.weight,
+                                     component.total_part_weight,
+                                     self._assignment.total_weight):
+
+                            # out of score for this part
+                            out_of_score = \
+                                Fraction(component.weight,
+                                         self._assignment.total_weight) * \
+                                Fraction(part.weight,
+                                         component.total_part_weight)
+                            component_out_of += out_of_score
+
+                            # total score for this part
+                            total_score = out_of_score * part_grade.score
+                            component_total += total_score
+                        else:
+                            total_score = Fraction(0, 1)
+                            out_of_score = Fraction(0, 1)
+
+                        # gets short name for class
+                        dot_idx = part.part.cls.rfind(".")
```

Comment:
probably should not be using `JUnitXMLTest` specific stuff here

i think below we can just say `part.part.description()`

Reply:
sure